src/functional/ffront/decorator.py

```diff
@@ -128,6 +128,23 @@ def is_cartesian_offset(o: FieldOffset):
     return deduced_grid_type if requested_grid_type is None else requested_grid_type
 
 
+def _field_constituents_shape_and_dims(
+    arg, arg_type: ct.FieldType | ct.ScalarType
+) -> Generator[tuple[tuple[int, ...], list[Dimension]]]:
+    if isinstance(arg_type, ct.TupleType):
+        for el, el_type in zip(arg, arg_type.types):
+            yield from _field_constituents_shape_and_dims(el, el_type)
+    elif isinstance(arg_type, ct.FieldType):
+        dims = type_info.extract_dims(arg_type)
+        if hasattr(arg, "shape"):
+            assert len(arg.shape) == len(dims)
+            yield (arg.shape, dims)
+        else:
+            yield (None, dims)
+    else:
+        raise ValueError("Expected a `FieldType` or `TupleType` thereof.")
```

Comment:
```suggestion
        raise ValueError("Expected `FieldType` or `TupleType`.")
```
This is more consistent with other error messages